Migrated the Ruby Linter to NodeJS Linter

Ajay-singh1 opened the pull request:
## Description

This PR replaces the Ruby linter to a modern NodeJS linter which is more flexible and advanced.

The linter passes all the docs test successfully and also it resolves https://github.com/istio/istio.io/issues/7906.

This PR fixes issue No. https://github.com/istio/istio.io/issues/15148

## Reviewers

<!-- To help us figure out who should review this PR, please 
     put an X in all the areas that this PR affects. -->

- [ ] Ambient
- [x] Docs
- [x] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure
- [ ] Localization/Translation

<!-- If this is a localization PR, please replace this line with the URL of the original English document. -->

Ajay-singh1 commented:
@craigbox The previous PR got cluttered up so I had to create a new one.Regarding the MD007 rule the ruby based linter doesnt strictly enforce the MD007 rule  and allows mixed indentation while the node based  linter enforces strict MD007 rule so a much better solution will be to disable the MD007 rule.

dhawton commented:
> @craigbox The previous PR got cluttered up so I had to create a new one.Regarding the MD007 rule the ruby based linter doesnt strictly enforce the MD007 rule and allows mixed indentation while the node based linter enforces strict MD007 rule so a much better solution will be to disable the MD007 rule.Once the check passes here I will update the common files so that the change is applied to all istio repos.The linting is sucessful now:-

I actually strongly disagree with disabling MD007. This should be something we enforced, it's been something I have been fixing in PRs that I review, it'd be much better to enforce and resolve the issues.

Ajay-singh1 commented:
@dhawton I need a couple of days to come up with a solution.Will be right back.

Arhell reviewed (commented):
/ok-to-test

istio-testing commented:
@Ajay-singh1: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 473a58d06b87b66e8fcceb0c2b019a48352b72db | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1899977238036615168) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | 473a58d06b87b66e8fcceb0c2b019a48352b72db | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1899977238036615168) | true | `/test lint`
gencheck_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1900398825609105408) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1900398825609105408) | true | `/test gencheck`
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900398825676214272) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

Ajay-singh1 commented:
/retest

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900398825676214272) | true | `/test lint`
gencheck_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1900403181247008768) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1900403181247008768) | true | `/test gencheck`
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900403181309923328) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

Ajay-singh1 commented:
@dhawton markdownlint-cli2 is not being recognized.

Ajay-singh1 commented:
/retest-required

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900403181309923328) | true | `/test lint`
gencheck_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1900427606507917312) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1900427606507917312) | true | `/test gencheck`
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900427606579220480) | true | `/test lint`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

Ajay-singh1 commented:
Complex Issue unable to resolve closing this.

craigbox commented:
@Ajay-singh1 are you intending to pick this up elsewhere?  I think you have everything you need to proceed, and I can help steward it.  Otherwise I can move it onto another maintainer.

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900427606579220480) | true | `/test lint`
gencheck_istio.io | af0164e8fc89aa120e9c9e114f05f085f9c2ef17 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1903979166852714496) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

Ajay-singh1 commented:
@craigbox Yes , this will work and will pass all the lint checks.I am waiting for the PR to be merged in istio/tools once it is approved the checks will pass here and we are good to go.

craigbox commented:
I don't see a PR.  Can you link one if it exists?  Else, do you want to add it to https://github.com/istio/tools/pull/3161 ?
(I would also add both for the meantime)

Ajay-singh1 commented:
Yes this specific PR I was talking about istio/tools#3161.It is in pending state once it merges the lint checks will pass here I am completely confident about it.

Ajay-singh1 commented:
/retest

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900427606579220480) | true | `/test lint`
gencheck_istio.io | af0164e8fc89aa120e9c9e114f05f085f9c2ef17 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1904450713933582336) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

craigbox commented:
You may have to wait a day or so for the automator to build a new container and update this repo to use it.

Ajay-singh1 commented:
👍

Ajay-singh1 commented:
/retest

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900427606579220480) | true | `/test lint`
gencheck_istio.io | af0164e8fc89aa120e9c9e114f05f085f9c2ef17 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1905101079985524736) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

craigbox commented:
Try pushing an empty commit. it will probably retest on an existing image
otherwise

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
lint_istio.io | a7288b4ae2464623bd8fb55d18c7c26123c86ebf | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/lint_istio.io/1900427606579220480) | true | `/test lint`
gencheck_istio.io | f6464eba12a01926060f85433200e611c1e55903 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1905126015965859840) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Ajay-singh1: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
gencheck_istio.io | 0c58b643710615e25412c88e75c738f0eac3ef95 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/gencheck_istio.io/1905128773964009472) | true | `/test gencheck`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

craigbox commented:
Nice!  The build is working.

The gencheck error is because you added a `<!-- markdownlint-disable -->` block to a code snippet, but not to the generated `snips.sh` file.  However, I would suggest we shouldn't have an ignore block anywhere but the front matter of a document.

I would also like to standardize on using a `markdownlint-disable: XXX` block in the doc front matter.  Can we see if this is natively supported or how we might need to introduce it?

craigbox commented on content/en/docs/ops/configuration/security/security-policy-examples/snips.sh:
```diff
@@ -81,6 +81,7 @@ spec:
   mtls:
     mode: STRICT
 ---
+<!-- markdownlint-disable-file MD007 -->
```
This content is now in the YAML files that are used for testing, which will almost certainly cause all the tests to fail.

Ajay-singh1 replied:
I agree with you I will fix this within a day.Yes markdownlint-disable works for doc front matter it is natively supported.

craigbox replied:
Yeah, that's confirmed here: https://github.com/DavidAnson/markdownlint/issues/1250

unfortunate.

istio-testing commented:
@Ajay-singh1: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
doc.test.profile-default_istio.io | e3e77466e559bf1bd40afe21bc393f6515787cd1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/doc.test.profile-default_istio.io/1905136884124749824) | true | `/test doc.test.profile-default`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@Ajay-singh1: The following tests **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:

Test name | Commit | Details | Required | Rerun command
--- | --- | --- | --- | ---
doc.test.profile-default_istio.io | e3e77466e559bf1bd40afe21bc393f6515787cd1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/doc.test.profile-default_istio.io/1905136884124749824) | true | `/test doc.test.profile-default`
doc.test.profile-none_istio.io | e3e77466e559bf1bd40afe21bc393f6515787cd1 | [link](https://prow.istio.io/view/gs/istio-prow/pr-logs/pull/istio_istio.io/16316/doc.test.profile-none_istio.io/1905136884418351104) | true | `/test doc.test.profile-none`

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes-sigs/prow](https://github.com/kubernetes-sigs/prow/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

craigbox commented on .markdownlint.json:
Why is this file both here and in `common/config`?  Can we have just the one and have it be copied?

Ajay-singh1 replied:
Agreed!

craigbox commented on .markdownlint.json:
```diff
@@ -0,0 +1,13 @@
+{
+    "default": false,
+    "MD002": { "level": 2 },
+    "MD007": { "indent": 4 },
```
This rule is disabled in many files.   Can you look into why they are failing, and if we can just change the markup to pass correctly?  Did we not have this rule in the old config?

Ajay-singh1 replied:
We did have this rule in the old configuration.It is failing because markdownlint-cli2 enforces strict MD007 rules while the previous linter allows for mixed indentation for the MD007 rule.Changing markup will have several breaking changes.

craigbox commented on .markdownlint.json:
```diff
@@ -0,0 +1,13 @@
+{
+    "default": false,
```
Does this turn off the default rules?  Wouldn't we want them on?

Ajay-singh1 replied:
Yes this will turn off the default rules , no turning them on is not a good way because all the rules will be applied by default which is used by the linter and we dont want them right?We will explicitly define the rules which we want to use to lint our markdown files.

craigbox commented on content/en/blog/2017/0.1-canary/index.md:
```diff
@@ -114,6 +114,8 @@ spec:
         subset: v2
       weight: 10
 ---
+<!-- markdownlint-disable-file MD007 MD026 -->
+<!-- markdownlint-disable-file MD007 -->
```
The first line should render the second line unnecessary

craigbox replied:
These comments. if they're truly needed, should only exist at the top of a file (directly under the front matter).  It looks like you might have done something like taking all the failing files and tried to add a comment under the second instance of `---`  ?

Ajay-singh1 replied:
Agreed! I am also of the same view actually this is the result of a bash script , it automatically added these lines in the files.

Ajay-singh1 commented:
I have addressed all the issues , We can cleanly merge this PR , If you have any further questions do let me know.Thanks @dhawton and @craigbox for your assistance.

Ajay-singh1 commented:
@craigbox Can I get a thumbs up here?
